### PR TITLE
db/hints: Log when ignoring invalid hint directories

### PR DESCRIPTION
**Pull request by dawmd**
In 58784cd, aa4b06a and other commits migrating hinted handoff from IPs to host IDs (scylladb/scylladb#15567), we started ignoring hint directories of invalid names, i.e. those that represent neither an IP address, nor a host ID. They remain on disk and are taken into account while computing, e.g. the total size of hints, but they're not used in any way.

These changes add logs informing the user when Scylla encounters such a directory.

### CODE REVIEW

**Comment by mykaul** [comment]
Please add tests.

**Comment by dawmd** [comment]
Tests verifying the changes: scylladb/scylla-dtest/pull/4006

Version 2 (I had to re-force-pushed the changes. I don't know why some other commits got involved):
* coroutinization of `space_watchdog::scan_one_ep_dir`
* taking invalid directories into account as well when computing the total space consumed by hint directories

File: db/hints/manager.cc [hunk 1]
```diff
@@ -185,15 +185,22 @@ future<> manager::start(shared_ptr<gms::gossiper> gossiper_ptr) {
 
     co_await lister::scan_dir(_hints_dir, lister::dir_entry_types::of<directory_entry_type::directory>(),
             [this] (fs::path datadir, directory_entry de) {
-        endpoint_id ep = endpoint_id{de.name};
+        const auto maybe_ep = std::invoke([&] () -> std::optional<endpoint_id> {
+            try {
+                return endpoint_id{de.name};
+            } catch (...) {
+                manager_logger.warn("Encountered a hint directory of invalid name: {}. It will be ignored.", de.name);
```
Comment (piodul) [inline db/hints/manager.cc:192]: > `"Encountered a hint directory of invalid name: {}. It will be ignored."`

I'd concentrate more on the consequences to the user, i.e. what it means that the directory will be ignored. In addition to the current message, I'd stress that hints from this directory will not be replayed.

File: db/hints/manager.cc [hunk 2]
```diff
@@ -185,15 +185,22 @@ future<> manager::start(shared_ptr<gms::gossiper> gossiper_ptr) {
...
+                return {};
+            }
+        });
```
[the remark this answers is not in the document]
Comment (piodul) [reply db/hints/manager.cc:195]: Ok, you convinced me that it's not necessarily a good idea. This is a tiny duplication that appears in a very small number of places, so we can live with it, and the alternative I proposed is not necessarily better, so let's keep it as it is.

**Comment by dawmd** [comment]
Version 3:
* split the changes into more but single-purpose commits
* made the logs more descriptive

File: db/hints/resource_manager.cc [hunk 3]
```diff
@@ -142,14 +146,25 @@ void space_watchdog::on_timer() {
                 // not hintable).
                 // If exists - let's take a file update lock so that files are not changed under our feet. Otherwise, simply
                 // continue to enumeration - there is no one to change them.
-                const internal::endpoint_id ep{de.name};
+                const auto maybe_ep = std::invoke([&] () -> std::optional<internal::endpoint_id> {
+                    try {
+                        return internal::endpoint_id{de.name};
+                    } catch (...) {
+                        resource_manager_logger.warn("Encountered a hint directory of invalid name: {}, exception: {}. ",
+                                de.name, std::current_exception());
```
Comment (piodul) [inline db/hints/resource_manager.cc:154]: AFAIK `space_watchdog::on_timer()` runs every second. This will generate a lot of spam. I don't think we should print it with such high verbosity, let's reduce it to `debug` - the user will know that there are issues with these directories because of the `warn` messages in `manager::start`.

Comment (dawmd) [reply db/hints/resource_manager.cc:154]: That's a good point, thank you

**Comment by dawmd** [comment]
Version 4:
* reduced the log level in `space_watchdog::on_timer()` to debug

**Comment by piodul** [comment]
@dawmd now that https://github.com/scylladb/scylladb/pull/15567 is merged, can this PR be closed? If so then please go ahead and close it.

**Comment by piodul** [comment]
> @dawmd now that #15567 is merged, can this PR be closed? If so then please go ahead and close it.

@dawmd ping

**Comment by dawmd** [comment]
Version 4:
* dropped most of the commits as the changes were implemented as part of scylladb/scylladb#15567
* rebased on top of the current master
* the remaining changes of the previous version of this PR are logs. We'll use them for testing too

**Comment by piodul** [comment]
@dawmd backport candidate label is not being honored by the "PR require backport label" check. You need to choose labels with specific versions (like e.g. `backport/6.0`) or `backport/none` if no backport is needed.

I'll add `backport/none` as this PR just adds more log messages.

File: db/hints/resource_manager.cc [hunk 4]
```diff
@@ -137,7 +137,8 @@ void space_watchdog::on_timer() {
         _total_size = 0;
         for (manager& shard_manager : per_device_limits.managers) {
             shard_manager.clear_eps_with_pending_hints();
-            lister::scan_dir(shard_manager.hints_dir(), lister::dir_entry_types::of<directory_entry_type::directory>(), [this, &shard_manager] (fs::path dir, directory_entry de) {
+            lister::scan_dir(shard_manager.hints_dir(), lister::dir_entry_types::of<directory_entry_type::directory>(),
+                    [this, &shard_manager] (fs::path dir, directory_entry de) {
```
Comment (piodul) [inline db/hints/resource_manager.cc:141]: This is a change that is completely unrelated to adding logs. I suggest dropping it.

**Review by piodul — COMMENTED** [review]
Just one nitpick, otherwise it LGTM

**Comment by dawmd** [comment]
Version 5:
* Dropped the unnecessary indentation change

**Review by piodul — APPROVED** [review]

**Comment by piodul** [comment]
Queued
